src/bdmsupport.c

```diff
@@ -314,38 +316,47 @@ static void bdmLaunchGame(int id, config_set_t *configSet)
     int irx_size = size_bdm_cdvdman_irx;
     compatmask = sbPrepare(game, configSet, irx_size, irx, &index);
     settings = (struct cdvdman_settings_bdm *)((u8 *)irx + index);
+    if (settings == NULL)
+        return;
+    memset(&settings->frags, 0, sizeof(bd_fraglist_t));
+
     for (i = 0; i < game->parts; i++) {
+        bd_fraglist_t part_frags;
+        int fidx;
+
+        // Open file
         sbCreatePath(game, partname, bdmPrefix, "/", i);
         fd = open(partname, O_RDONLY);
-        if (fd >= 0) {
-            int *pBDMDriver = (int *)bdmDriver;
-            *pBDMDriver = fileXioIoctl(fd, USBMASS_IOCTL_GET_DRIVERNAME, "");
-            LOG("bdmDriver=%s\n", bdmDriver);
-
-            if (settings != NULL)
-                settings->LBAs[i] = fileXioIoctl(fd, USBMASS_IOCTL_GET_LBA, "");
-
-            if (fileXioIoctl(fd, USBMASS_IOCTL_CHECK_CHAIN, "") != 1) {
-                close(fd);
-                // Game is fragmented. Do not continue.
-                if (settings != NULL)
-                    sbUnprepare(&settings->common);
-
-                guiMsgBox(_l(_STR_ERR_FRAGMENTED), 0, NULL);
-                return;
-            }
+        if (fd < 0) {
+            sbUnprepare(&settings->common);
+            guiMsgBox(_l(_STR_ERR_FILE_INVALID), 0, NULL);
+            return;
+        }
 
-            if ((gPS2Logo) && (i == 0))
-                EnablePS2Logo = CheckPS2Logo(fd, 0);
+        // Get driver - we should only need to do this once
+        int *pBDMDriver = (int *)bdmDriver;
+        *pBDMDriver = fileXioIoctl(fd, USBMASS_IOCTL_GET_DRIVERNAME, "");
 
+        // Get fragment list
+        fileXioIoctl2(fd, USBMASS_IOCTL_GET_FRAGLIST, NULL, 0, (void *)&part_frags, sizeof(bd_fraglist_t));
+        if ((settings->frags.count + part_frags.count) > 10) {
```

Comment:
maybe 10 can be delivered from header? just in case we increase the limit so do not forget to increase it here too.

Reply:
Good point. But unfortunately I didn't add it to the header in ps2sdk :(. There's also another thing bothering me. Instead of 10 fragments it would be better to be able to ask for more (or dynamic) number of fragments.

Having more fragments could also make this new structure compatible with HDD fragments:

https://github.com/ps2homebrew/Open-PS2-Loader/blob/08fc57b2f758ffcfd8b1917a1df96cb23c20b7fb/modules/iopcore/cdvdman/atad.h#L77-L84

Perhaps the best solution is to make the function (`USBMASS_IOCTL_GET_FRAGLIST`) dynamic, and pass a fixed amount of 65 fragments to the iopcore? Then as a second step, make the HDD and BDM fragments use the same structure and code?